graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsSimpleDataLoaderInstrumentation.kt

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2024 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.graphql.dgs
+
+import org.dataloader.BatchLoaderEnvironment
+
+/**
+ * Provides a simple interface to define instrumentation around the dispatch of data loaders.
+ *
+ * Requires [com.netflix.graphql.dgs.internal.DgsWrapWithContextDataLoaderScanningInterceptor] to be enabled,
+ * as DgsSimpleDataLoaderInstrumentation only provides hooks for the WithContext versions of batch loaders.
+ */
+interface DgsSimpleDataLoaderInstrumentation<C> {
```

Comment:
Does the generic type actually add any value here? I think it would practically be cleaner to just return / accept a context parameter as `Any`, and then you wouldn't need the `afterLoadWithCast` method. It seems like it's always a star-projected type so it's always `Any?` in practice.

Reply:
It allows someone using the interface to have type safety in their implementation. I was (and am) pretty torn on whether or not it's worthwhile though.

Reply:
The `InstruementationState` example is analogous to what I am suggesting though. `Instrumentation` doesn't have an generic type, all of the methods take an `InstrumentationState` instance and the implementation needs to cast to the specific type.

I guess another approach is to change the flow a bit and have the returned context have the callback method for the "onComplete" event; it would be more analogous to the `Instrumentation` stuff in graphql-java already.

```
interface BatchLoaderInstrumentation {
    fun beginLoad(context: LoaderContext): BatchLoaderInstrumentationContext
}

interface BatchLoaderInstrumentationContext {
    fun onDispatched(result: CompletableFuture<*>)
    fun onCompleted(result: Any?, t: Throwable?)
}
```

Then we don't really have to deal with unchecked casts and the like.

Reply:
Actually based on [this](https://github.com/graphql-java/graphql-java/pull/3457), the interface could be simplified.